brushfire-core/src/main/scala/com/stripe/brushfire/local/Trainer.scala

```diff
@@ -3,81 +3,176 @@ package com.stripe.brushfire.local
 import com.stripe.brushfire._
 import com.twitter.algebird._
 
-case class Trainer[K: Ordering, V, T: Monoid](
-    trainingData: Iterable[Instance[K, V, T]],
-    sampler: Sampler[K],
-    trees: List[Tree[K, V, T]]) {
 
-  private def updateTrees(fn: (Tree[K, V, T], Int, Map[LeafNode[K, V, T, Unit], Iterable[Instance[K, V, T]]]) => Tree[K, V, T]): Trainer[K, V, T] = {
-    val newTrees = trees.zipWithIndex.par.map {
-      case (tree, index) =>
-        val byLeaf =
-          trainingData.flatMap { instance =>
-            val repeats = sampler.timesInTrainingSet(instance.id, instance.timestamp, index)
-            if (repeats > 0) {
-              tree.leafFor(instance.features).map { leaf =>
-                1.to(repeats).toList.map { i => (instance, leaf) }
-              }.getOrElse(Nil)
-            } else {
-              Nil
-            }
-          }.groupBy { _._2 }
-            .mapValues { _.map { _._1 } }
-        fn(tree, index, byLeaf)
-    }
-    copy(trees = newTrees.toList)
+trait TrainingStep[K,V,T] {
+
+  type K1
+  type V1
+  type V2
+
+  def prepare(trees: Map[Int, Tree[K,V,T]], instance: Instance[K,V,T]): Seq[((Int,Int,K1), V1)]
+  def lift(tree: Tree[K,V,T], leafIndex: Int, key: K1, v1: V1): Traversable[V2]
+  def update(tree: Tree[K,V,T], map: Map[Int,V2]): Tree[K,V,T]
+
+  def semigroup1: Semigroup[V1]
+  def semigroup2: Semigroup[V2]
+  def ordering: Ordering[K1]
```

Comment:
So, I was thinking about this case, and I think it may make more sense to have our general abstraction be something like,

```scala
trait GenTrainer[K, V, T] {
  type TrainingStepContext[K1, V1, V2]

  def trainingStep(step: TrainingStep[K,V,T])(implicit ctx: TrainingStepContext[step.K1, step.V1, step.V2]): Trainer[K,V,T] = ???
}
```

(Or something similar.) In the local case this would basically be a no-op, but for Scalding we may want to capture and `Ordering` and for Spark maybe something else (`ClassTag`s? ¯\_(ツ)_/¯)